command/agent/consul/check_watcher.go

```diff
@@ -103,20 +103,34 @@ func (c *checkRestart) apply(ctx context.Context, now time.Time, status string)
 		c.logger.Debug("restarting due to unhealthy check")
 
 		// Tell TaskRunner to restart due to failure
-		const failure = true
 		reason := fmt.Sprintf("healthcheck: check %q unhealthy", c.checkName)
 		event := structs.NewTaskEvent(structs.TaskRestartSignal).SetRestartReason(reason)
-		err := c.task.Restart(ctx, event, failure)
-		if err != nil {
-			// Error restarting
-			return false
-		}
+		go asyncRestart(ctx, c.logger, c.task, event)
```

Comment:
Question: does it make sense to use a semaphore or channel blocking technique used elsewhere, so we don't call `task.Restart` concurrently and if we get a spike of Restarts applies, we only restart once?

Reply:
Good question!

The checkWatcher.Run loop removes a check after Restart is called, so the same task won't be restarted more than once (until it completes the restart and re-registers the check).

In 0.8 TR.Restart just ticked a chan and so was async without having to create a new goroutine. This seemed like the least risky way of replicating that behavior.

Tasks that fail in a tight loop ([`check_restart.grace=0`](https://www.nomadproject.io/docs/job-specification/check_restart.html#grace) and [`restart.delay=0`](https://www.nomadproject.io/docs/job-specification/restart.html#delay)) could in theory spin up lots of goroutines, but the goroutines for a single task should rarely if ever overlap and restarting a task already involves creating a lot of resources more expensive than a goroutine.

That being said I hate this "fire and forget" pattern, so I'm open to ideas as long as they can't block `checkWatcher.Run`/`checkRestart.apply`. (`checkWatcher` should probably be refactored to separate Watch/Unwatch mutations from check watching, but that seemed way too big a risk for a point release)